snappy/snap_remote_repo.go

```diff
@@ -383,3 +386,61 @@ func (s *SnapUbuntuStoreRepository) Updates() (parts []Part, err error) {
 func (s *SnapUbuntuStoreRepository) Installed() (parts []Part, err error) {
 	return nil, err
 }
+
+// Download downloads the given snap and returns the filename of the local
+// version
+func (s *SnapUbuntuStoreRepository) Download(remoteSnap *RemoteSnapPart, pbar progress.Meter) (string, error) {
+	w, err := ioutil.TempFile("", remoteSnap.pkg.Name)
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		if err != nil {
+			os.Remove(w.Name())
+		}
+	}()
+	defer w.Close()
+
+	// try anonymous download first and fallback to authenticated
+	url := remoteSnap.pkg.AnonDownloadURL
+	if url == "" {
+		url = remoteSnap.pkg.DownloadURL
+	}
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return "", err
+	}
+	setUbuntuStoreHeaders(req)
+
+	if err := download(remoteSnap.Name(), w, req, pbar); err != nil {
+		return "", err
+	}
+
+	return w.Name(), w.Sync()
```

Comment:
syncing does not do the whole dance. Unfortunately the atomic file operation we have takes a []byte, instead of implementing a writer. Maybe we should refactor it a bit and then use that here; having a WriterComitter interface kind.